### ISSUE TITLE
Add the ability to enable Backdrop-only required modules during an upgrade from Drupal 7.

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5499]
There are modules (e.g., Entity Plus, Entity UI) that didn't exist in D7 and are required by Backdrop versions of D7 modules. Examples of B modules that need Entity Plus (E+, for short), for example, are Rules, Ubercart, and [the modules in this list](https://github.com/backdrop-contrib/rules/issues/138#issuecomment-1032112517).

A problem arises when upgrading a D7 site that uses those modules. If the user merely adds the E+ module to the B codebase (the logical thing to do), when trying to upgrade the site, B sees that there was no enabled version of the D7 module, so it assumes the B version is not enabled, and therefore disables it and all of the downstream modules (like Rules, Ubercart, and their submodules), and so doesn't run their update hooks, and the user ends up with a site with only partially upgraded and a bunch of disabled modules. (This is particularly problematic with Ubercart, which has about 20 submodules, plus further contrib sub-submodules.)

The current solution, which has been recommended for Rules and Ubercart for a while, has been _stub modules_, basically fake/empty D7 modules that you enable before performing the upgrade. See [here](https://github.com/backdrop-contrib/rules/wiki/Upgrading-Rules-from-Drupal-7) and [here](https://github.com/backdrop-contrib/ubercart/wiki/Upgrading-Ubercart-from-Drupal-7).

More recently, [it was suggested](https://github.com/backdrop-contrib/rules/issues/138) to enable the B modules in an update hook. We can't actually do this during `hook_update_N()`, because by the time that would get called, the damage is done: the downstream modules are already disabled, so their update hooks don't get called. However, it can be done during `hook_requirements('update')` (suggested example [here](https://github.com/backdrop-contrib/manualcrop/commit/033d6c4cf80186292fbf6089e1f6b84c15ce4fbd), PR that does this for Rules module [here](https://github.com/backdrop-contrib/rules/pull/172)). So that works, but it's not ideal, for a couple of reasons:

* If this is the preferred solution, (nearly) the same chunk of code would need to be implemented by each contrib module that requires E+, E-UI, etc.
* More importantly, if we enable the required module by calling `module_enable()` as part of `hook_requirements()` (as in the PR), this results in calling `hook_enable()` for the module, and `hook_modules_enabled()` for _all other modules_, and this happens before the whole series of `hook_update_N()` calls has happened; so these hooks, which might be expecting fully-configured B modules, will be fired before the B modules are upgraded from D7.
* Also, if multiple modules are doing this, there's no particular control over the order in which the modules are enabled. (If that matters.)

So, it would be nice if Backdrop core was able to provide support for this situation: **when an enabled D7 module is being upgraded to B and depends on a B module that didn't exist in D7**.

Some possibilities:

* Add a `module_enable_lite()` function that lets the client module "enable" the dependency B-only module (i.e., create the appropriate entry in the `system` table) without firing the enable hooks, for use by the contrib modules in their `hook_requirements()` call.
* Add a new updating hook that lets client modules declare dependency B-only modules that ought to be treated as enabled if they exist in the codebase, and then core handles the "lite" enablement of the B-only modules.

We could also just put E+ (which is currently the most popular contrib module) and E-UI into core, so that they're already enabled in new installs. (But this wouldn't be a total solution; B Ubercart still has a B-only module that isn't in D7.)

This issue arose from discussions and issues with @argiepiano and @laryn; please add anything I've missed, and everyone else, please give your thoughts. (H/T also to @indigoxela, who raised [the issue for Rules](https://github.com/backdrop-contrib/rules/issues/138).)

---

Advocate: @bugfolder
PR: https://github.com/backdrop/backdrop/pull/3950

**Comment by docwilmot** [comment]
> More importantly, if we enable the required module by calling module_enable() as part of hook_requirements() (as in the PR), this results in calling hook_enable() for the module, and hook_modules_enabled() for all other modules, and this happens before the whole series of hook_update_N() calls has happened; so these hooks, which might be expecting fully-configured B modules, will be fired before the B modules are upgraded from D7.

I'm not seeing quite why this is a problem though. Both `hook_enable()` and `hook_modules_enabled()` in that case would only be passed the `entity_plus` and `entity_ui` modules right, so unless those two actually implement those hooks, then nothing "bad" should happen?

Thinking about update.php though, there should never normally be a time when a module is running updates and finds that it newly depends on another, disabled, module. This should reasonably only happen during a B-7 upgrade. So I think its reasonable to put a little stop in update.php when its checking for update dependencies, so that it pauses and asks "module x depends on module y, which is disabled, do you want to enable it?", or even a checkbox thats asks "Enable dependencies?". (I think it should be a choice rather than auto). We can then have update.php simply as possible switch the disabled module's status to 1 in System's tables, instead of fully calling `module_enable()` to avoid whatever complications you have found arise from that.

**Comment by argiepiano** [comment]
@bugfolder, thanks for this very thorough summary. 

@docwilmot, the idea of asking the user may be a bit confusing. Basically, there is no option. Either you enable E+, or you can't use Rules or Ubercart in your migrated site (or even worse, you'll get undefined functions all over the place). If a user is migrating a site with Rules or Ubercart, it's because they want the site to be as functional as possible once migrated.

So, another option, which would prevent unintended enabled modules, is to add a key to the Backdrop's Rules and Ubercart info file - something like `enable_on_migration` thus, Backdrop's Rules would have 
```
enable_on_migration[] = entity_plus
enable_on_migration[] = entity_ui
```

This way we are VERY specifically telling the update script that we want those (and only those) modules enabled when migrating, and only then.

Regarding your comment:
> We can then have update.php simply as possible switch the disabled module's status to 1 in System's tables, instead of fully calling module_enable() to avoid whatever complications you have found arise from that. 

This is tricky. It will work with Entity Plus and Entity UI because those two modules do not create tables in the DB. But, I can see a situation when we need to enable a module that DOES need to create a table - in which case switching the entry to 1 in the db will not work.

**Comment by docwilmot** [comment]
> the idea of asking the user may be a bit confusing

I understand. The issue here though is this will apply in any future situation where a user runs an update and a module requires a disabled module, which would otherwise be automatically enabled if we dont ask. I again confess I cant think why this would ever happen, but I would prefer if it did the user be warned. If not a checkbox, at least something like a message "module E will be enabled, since modules X . Y, Z depend on E" with a "continue" button or such might make sense. Then there will be no real option.

> This is tricky. 

Yes this is the "simply as possible" bit. I mean, just extract the minimal from `module_enable()`: if tables, install them, if none just whats necessary to switch to enabled without running all hooks.

> enable_on_migration[] = entity_ui

Does Backdrop know when an update.php run is a migration? Or isnt it all just `update_Ns` running in reality? This would still be special-casing a D7-B migration though, whereas looking for dependencies would be a more generic solution I feel.

**Comment by docwilmot** [comment]
Good catch. I'm still feeling that catching dependencies would be preferable to a new `enable_on_migration` key (which would have no other purpose apart from first upgrade yet persist forever in the info file, and which would also duplicate the `dependencies` key in the info file). Furthermore, for users who dont recognize they need E+ for example, and dont place it in the modules folder, they would still end with the same issue of non-enabled Rules etc, so it would be preferable to have a pause in the update process for a message that says required modules are missing.

**Comment by bugfolder** [comment]
How about instead of putting more info in the .info file (which, as noted, persists _visibly_ forever) if we add a hook, like `hook_declare_upgrade_required_modules($modules)` in the install file, where a module could declare B modules that should be lite-enabled? Then we wouldn't need to ask the user anything, and while that code would persist forever, like `hook_update_N()` calls, it can live in `mymodule.install` with the other functions that are only needed once in the lifetime of the site.

**Comment by argiepiano** [comment]
> Furthermore, for users who dont recognize they need E+ for example, and dont place it in the modules folder, they would still end with the same issue of non-enabled Rules etc, so it would be preferable to have a pause in the update process for a message that says required modules are missing.

Actually, the update script currently stops the process if Entity Plus (or any other dependency declared in the info file) is not present in the modules folder.  The issue we are trying to catch is when the E+ is present, but the database doesn't have it as enabled (because it did not exist in D7).

> How about instead of putting more info in the .info file (which, as noted, persists visibly forever) if we add a hook, like hook_declare_upgrade_required_modules($modules) in the install file

That seems like a good strategy. I'm going to try putting something together along these lines tonight.

**Comment by argiepiano** [comment]
I have created a PR (https://github.com/backdrop/backdrop/pull/3950) as a first step. This PR:

1. Only runs during the migration process, on the very first update.php of a D7 database 
2. Gets all enabled modules from the database
3. Invokes `hook_migrate_required_modules_info()` - a new hook that must be placed in the install file
4. Gathers a list of all modules to be enabled. If they are not present in the modules folder, nothing is done. update.php will stop the process if that's the case, later in the process
5. Invokes `update_module_enable()` with those modules. This function is a "lite enable" which won't fire any "enable" or "install" hooks

I've tested with a module that depends on Entity Plus (Entity Tokens), and it worked well. Entity Plus was enabled at the end of the process (NB: no "stub" modules was used in D7, AND Entity Plus was present in the modules folder).

Please test with Rules and others.

`hook_migrate_required_modules_info()` must return an array of modules needed to be enabled.

```php
/**
 * Implements hook_migrate_required_modules_info().
 */
function entity_token_migrate_required_modules_info() {
  return array('entity_plus');
}
```

One option I did not test yet is - what happens if the "stub" Entity Plus was enabled in the D7 db? I'll test that tomorrow.

**Comment by argiepiano** [comment]
I've modified the PR a bit to remove duplicate modules from the array of modules to be enabled, and to remove modules that are already enabled.

This takes care of my comment re: stub modules in the previous comment, and makes the enable process more efficient by passing an array with unique values.

**Comment by klonos** [comment]
> This would still be special-casing a D7-B migration though...

I don't think that this argument should be a show-stopper for any suggested solution in this thread. It is safe to assume that if sites are to upgrade from D7, then that will happen in the 1.x cycle of Backdrop. If we don't want this special-casing, we can remove it in Backdrop 2.0. Right?

> ...so it would be preferable to have a pause in the update process for a message that says required modules are missing.

I agree that that would be useful 👍🏼 

> Actually, the update script currently stops the process if Entity Plus (or any other dependency declared in the info file) is not present in the modules folder.  ...

That's great then! ...I could test things and find out myself, but since you already mentioned it @argiepiano can you please let me know if this is a "graceful" pause, with a helpful message, or if it is a php error/exception or otherwise unhelpful message? If you have a screenshot of what happens at that point, then that would also be great.

...and a HUGE thanks for putting a PR together already! 🙏🏼 ❤️

**Comment by docwilmot** [comment]
A hook works better, but still feels unnecessary IMO. Backdrop already knows that module X requires module Y: its in the info file, as `dependencies[]`. Adding a hook, though less visible I guess than another info file declaration, still duplicates. And I feel that any situation where this occurs would then benefit from a dialog, even if the disabling was accidental or bug or a D7-B upgrade, to let the user knows he's got a missing dependency.

**Comment by argiepiano** [comment]
> A hook works better, but still feels unnecessary IMO. Backdrop already knows that module X requires module Y: its in the info file, as dependencies[]. Adding a hook, though less visible I guess than another info file declaration, still duplicates.

I can see your point, @docwilmot. To me this hook feels like a safety measure - the developer is explicitly telling update.php that indeed, we need to enable a module that did not exist with that name in D7. 

In a typical migration situation most of the modules listed in `dependencies[]` are modules that existed in D7 with the same name in D7 and are already enabled in the D7 database. Therefore checking if they are enabled is a bit of a waste of processing time, and is unnecessary for most of them. (the script already checks if those modules are present in the folder).  

Any thoughts, @bugfolder and @klonos?

Re: adding  a dialog. That will require a very different approach, and I may need help with this. The question is: how do we force-stop the process with a form? And then, once submitted, how do we restart the process where it was stopped? We may need to make further modifications to the update.php. Thoughts?

**Comment by argiepiano** [comment]
> In our use case here though, there is something that we can do to fix the issues in this screen (but there is no submit button to indicate and allow that)

When you get this error message in the current version of update.php you are supposed to go back and put the missing modules in the modules folder, then restart (the bottom of this says "Check the messages and try again"

In the situation we are trying to fix with this PR there is basically nothing you can do to fix the problem. If you put a dialog saying something like "The migration script needs to enable Entity Plus to make your migration functional. Do you wish to proceed?"  If you say NO, basically you have a non-functional migration!

**Comment by argiepiano** [comment]
You can look at it this way:

In a normal Backdrop or D7 site, if you have Rules enabled, you simply can't disable Entity API or Entity Plus. The UI does not give you that option. Its checkbox is grayed out. There is a reason for this. You basically have no option. You MUST have Entity Plus or Entity API enable.

This is a similar situation, except that during the migration process Entity API is not called Entity API in Backdrop, rather, Entity Plus. So by automatically enabling it (and not giving them the option) we are not taking away a freedom that they had before. They never had the "freedom" to say no to Entity Plus.  If you want Rules, you are required to have Entity Plus/Entity API

**Comment by argiepiano** [comment]
After further thought: give me a day or two and I'll try to put something together that:
1. Has a way to inform the user that this enabling is happening
2. Uses dependencies[] instead of the new hook

I do think those options are really worth considering and I appreciate your suggestions, @klonos and @docwilmot

**Comment by bugfolder** [comment]
> Adding a hook, though less visible I guess than another info file declaration, still duplicates.

Not entirely. The idea of the hook was that this was a way of declaring module dependencies that could be _automatically_ enabled if the module code was present, as opposed to the default behavior of not auto-enabling non-enabled upstream modules (and therefore disabling the downstream modules).

But I am inclined to agree with @argiepiano's argument that in proper use, the user would have no choice about having E+ enabled, and that doing a D7->B upgrade might warrant special rules to apply. So I think it would be OK to just let the user know that any modules that are required by `dependencies[]` of enabled modules and exist in the codebase are going to be enabled automatically (so no extra hook required).

Looking forward to that next/updated PR.

**Comment by argiepiano** [comment]
While working on this issue I discovered that there are 8 core modules (all of them declared as dependencies of `standard.install`) that are never enabled during the migration. See #5506. The PR I'm creating would solve that issue too.

**Comment by argiepiano** [comment]
I have updated the PR

- There is an informational message that tells the user that some modules had to be enabled. This includes the 8 core required modules that were not enabled during migration - see #5506 
- The disabled required modules are all enabled using `module_enable()` rather than the "lighter" `update_module_enable()`. This was necessary because some of the modules need to create tables in the DB
- I have tested with a small module that depends on entity plus, and things seem to work fine. However, testing is needed with more complex modules like Rules and Ubercart
- This PR does not contain any type of checkbox or confirmation process like the ones mentioned above and in #5506. I'm still unsure this is necessary. These "core 8" and Entity Plus are ALL declared as dependencies of other modules, so it's safe to assume that they are  needed by a functional installation, and giving the option of "not enabling" them will lead to problems (otherwise they would not be declared as dependencies!). This PR, however, does inform the user of the enabling.

<img width="848" alt="Screen Shot 2022-02-11 at 9 57 59 PM" src="https://user-images.githubusercontent.com/9938978/153698105-e955e990-224a-4409-9600-3ebdb72e92da.png">

**Comment by klonos** [comment]
Thank you for working on this @argiepiano 🙏🏼 

Can we please make the modules in the message a proper `<ul>` list (using `theme_item_list()`), and have the human-readable names of the modules there?

**Comment by argiepiano** [comment]
Thanks, @klonos. The current PR does that (I'm aware of your suggestions in terms of formatting the notice with bullets - I'll get to that soon).   I also need to take a look at the failing tests with this PR to see what's going on. 

I am also interested in creating an alternative PR that shows a form with checkboxes, which is more or less what @docwilmot  suggested above.  I hope to work on this over the weekend.

**Comment by argiepiano** [comment]
I have modified the PR  as follows:
- The user is informed of modules that will be enabled BEFORE they are enabled. This is included together with the "info" screen that is shown before the updates
- The modules are enabled right before the list of updates is shown, before the user proceeds with the update

As before, the automatic enabling ONLY happens if we are migrating from D7.

Regarding the failing tests: I've looked briefly at some of them. Some are happening because, in order to check if modules need to be enabled, the new code invokes `system_rebuild_module_data()`. The initial problem is that, during the tests (and ONLY during the tests), some needed inc files have not been loaded.  I can try a few different things this week, but I'll appreciate suggestions.

To test: create a barebones D7 site, enable Rules and Entity API, and then migrate it to Backdrop.

**Comment by argiepiano** [comment]
This is a partial screenshot of the "info" screen with the list of modules needed to be enabled.

<img width="706" alt="Screen Shot 2022-02-20 at 11 03 32 PM" src="https://user-images.githubusercontent.com/9938978/154898220-c0265c50-221a-4a13-a785-26db5af811e4.png">

**Comment by bugfolder** [comment]
I set up a D7 installation with Rules enabled, then set up a Backdrop upgrade using the core code from the PR. The upgrade proceeded successfully, and at the end of it, Rules, Entity UI, Entity Plus, and Entity tokens were all enabled. But not all modules were listed as being enabled. Here's what that screen showed. No "Entity Plus", etc.:

![enabled](https://user-images.githubusercontent.com/29214084/155053832-8dc750aa-6cb2-49dc-9dd7-a89e9009ce82.jpg)

Also, I would suggest this wording for that line below "Some modules need to be enabled":

The following Backdrop modules will be enabled in order for the site to be upgraded successfully:

I.e., "will be," not "need to be," and "upgraded," not "migrated".

I also want to try the same thing with a full Ubercart, but that will have to be tomorrow.

**Comment by argiepiano** [comment]
@bugfolder thank you for testing. This is very strange. Are you sure that your D7 site didn't have the "stub" modules for Entity Plus and Entity UI enabled before the migration? The code actually loops through all enabled modules in the D7 database (including Rules), and checks that the required modules for those are also enabled. The only case in which Entity Plus would not be shown would be if the stub Entity Plus was enabled in the D7 site.

I'll double-check on this tonight or tomorrow.

Thanks for the wording suggestions.

**Comment by bugfolder** [comment]
Thought I didn't have them present, but I'll double-check tomorrow.

**Comment by bugfolder** [comment]
Oh that was interesting! Although I had disabled and uninstalled the stub modules in my Drupal->Backdrop test rig before starting the upgrade, they still had entries in the `system` table (listed as uninstalled), where they were set to uninstalled. But the upgrade still worked, properly enabling E+, EUI, and ET.

I also did a full Ubercart upgrade this morning, and that, too, worked (but interestingly, the Ubercart module showed up in the list of modules to be enabled.)

I'll do one more test today starting from a clean Drupal install with no stub modules anywhere in sight, but I'm gonna guess that it'll work just fine.

**Comment by argiepiano** [comment]
Thanks @bugfolder. 

Regarding your first paragraph: did the upgrade script still not show E+, EUI and ET in the list of modules needed to be enabled?

Regarding Ubercart: it's most likely shown there because the D7 version did not have a `ubercart.module` like the Backdrop version does. In [D7 Ubercart](https://git.drupalcode.org/project/ubercart) you have to enable the modules marked as REQUIRED one by one to be able to use Ubercart (uc_store, uc_order, uc_cart and uc_product). Whe Ubercart was ported to Backdrop, someone created a single module called ubercart.module (empty) which enables those four modules.

So, yes, since D7 database did not list ubercart.module a enabled, the Backdrop update script will show it as "needed to be enabled"

**Comment by bugfolder** [comment]
> did the upgrade script still not show E+, EUI and ET in the list of modules needed to be enabled?

Nope. And now I've run an upgrade from a completely fresh install of Drupal with just these modules:

![modules](https://user-images.githubusercontent.com/29214084/155206995-177c1c94-235d-4f64-80c3-fd75b449d16e.jpg)

And I confirmed that prior to starting the upgrade, the system table did not contain any entries for E+, EUI, or ET.

Nonetheless, here's what the upgrade screen showed:

![enabled](https://user-images.githubusercontent.com/29214084/155207138-21ea6aeb-ad7e-4fe7-bd16-cad3b4ab2734.jpg)

The upgrade worked, Rules, E+ and EUI enabled, and Rules ran its update hooks. So, the upgrade worked. But still no listing of the new modules.

**Comment by argiepiano** [comment]
@bugfolder I've looked at why Entity Plus is not listed in the modules to be enabled. 

Here's a question: are you perhaps using a version of Rules patched with the [PR 172](https://github.com/backdrop-contrib/rules/pull/172)  of Rules? If so, that patch includes an implementation of hook_requirements() that enables Entity Plus before the new code in update.php gets to it. Can you please check?

**Comment by bugfolder** [comment]
Yup, that's it. The .info file says it's 1.x-2.2.0, but the .install file contains `rules_requirements()` with that patched code, so yeah, I must have tested the patched version in the past and then forgot I'd done that and so re-used the module. (Pretty good sleuthing there!) Will run again with a fresh download of 1.x-2.2.0.

**Comment by bugfolder** [comment]
Success:

![enabled](https://user-images.githubusercontent.com/29214084/155407116-9eda052d-badc-4581-a1e5-9677f43097b4.jpg)

And we'd previously established that it works for Ubercart, so, I'd call this LGTM. But maybe there's other scenarios that should be tested?

**Comment by argiepiano** [comment]
Thanks @bugfolder.  We still need to make sure the tests don't break. I'm hoping to get to that soon.

**Comment by olafgrabienski** [comment]
> so, I'd call this LGTM

Reminder regarding the wording, so that it doesn't get lost:

> Also, I would suggest this wording for that line below "Some modules need to be enabled":
> The following Backdrop modules will be enabled in order for the site to be upgraded successfully:
> I.e., "will be," not "need to be," and "upgraded," not "migrated".

**Comment by argiepiano** [comment]
Yes, thanks @olafgrabienski. I've already made that change and it will be included in the next PR.

**Comment by argiepiano** [comment]
The tests are now passing 🎉 

At the risk of repeating some of the things that @bugfolder said in the OP, I feel I need to explain what this PR does. I'd like some feedback, as I feel a bit uneasy about a couple of things below.

What this PR does
------------------
In a nutshell, it gets all the enabled modules from the D7 database, and checks to see if their Backdrop counterpart modules have dependencies that are not already enabled in the D7 database. If it finds any required modules that need to be enabled, it enables them after notifying the user in the first info page in the update process. This check and auto-enabling only happens when upgrading a D7 site to Backdrop - not during a "normal" update.php run.

The typical scenario: when upgrading a D7 with modules like Rules or Ubercart, that use Entity API as a dependency.  The functionality of the original D7 Entity API module is now "spread out" in several different modules in Backdrop:  core's Entity, and the contrib Entity Plus, Entity UI and Entity Tokens. Entity Plus and Entity UI did not exist in D7, and therefore they would need enabled in order for the site to work once upgraded.

In addition, there are a few Backdrop core modules that are part of the Standard Backdrop installation, that did not exist in D7's core (Redirect, Installer, Email, Link, CKEditor, etc). This PR will also automatically enable those modules when upgrading a Standard D7 installation.

How the enabling is done
------------------------
Because modules like Redirect, Email and others must create either tables in the DB, or config files, the only possible way to enable them automatically is to use `module_enable()`, instead of the "lighter" `update_module_enable()`, which is typically used in update processes.  

This is one of the things I feel uneasy about. I don't think there is a way to avoid `module_enable()` at this point. As far as I can tell through my tests and through the automated tests, using `module_enable()` is not causing any issues... but I'd like thoughts about this.

Why the tests were failing: core's Entity module not available early on.
----------------------------------------------
The automated upgrade tests simulate three upgrade process from a D7 site to a Backdrop site.  During those upgrades, this PR invokes `module_enable()` to enable those core modules I mentioned above (Redirect, Installer, etc). The issue with this is that `module_enable()` uses a bunch of functions provided by core's entity.module. The problem is that, in a barebones D7 site, there isn't an `entity.module`,  and therefore those function invocations fail.   

My workaround is to actually enable Backdrop core's entity.module early in the process, ONLY if we are dealing with a D7 upgrade. The enabling, again, has to be done with `module_enable()`, in order to load the module and have those needed functions available.

----
To recap:  what are your thoughts about using `module_enable()` early in the upgrade process?  One way of avoiding `module_enable()` is to ONLY enable on Entity Plus and Entity UI (which don't need any tables or config created and can use the "light enable" with update_module_enable()) and to avoid enabling those core modules (Redirect, Installer, etc) I mentioned above.

**Comment by docwilmot** [comment]
I was going to suggest we modify `update_module_enable()` to check for schema and install. But I see this line already there that I dont understand: `$function = $module . '_schema_0'`. What is that?

**Comment by argiepiano** [comment]
Yeah, from what I've been able to gather, `update_module_enable()` and `hook_schema_0()` is a 10-year old relic from the initial development of D8, and it was there to help the upgrade from D7 into D8. 

That function doesn't exist anymore in Drupal (in any version). You can see [some remaining issues that address it](https://www.drupal.org/project/drupal/issues/1239370#comment-6505768)

So, from that discussion, I believe `hook_schema_0()` was used so that individual (core and/or contrib) modules could specify a "base schema" to be created at the very beginning of the upgrade process before any `hook_update_` equal or greater than 8000 could be run.

The idea of installing a module's db schema (and config file) in upgrade from D7 to Backdrop is valid and important, as we are seeing in this PR (e.g. when enabling Redirect during the upgrade). 

So, one possible solution to the issue at hand here is to implement hook_schema_0() for those Backdrop core modules that enabled by this PR. That would allow for the creation of the required tables.

In addition to that, we also need to ensure that config files are created, if we are not going to call module_enable() during this process.

**Comment by argiepiano** [comment]
I'd like to get the conversation going here to see what we need to do next. 

As I see it, we have two options:

1. Use `module_enable()` to enable the required modules, like in the latest PR. These calls happen early in the upgrade process (which may be a cause for concern), but ONLY when upgrading from D7, not when updating. In my several tests, and in the provided automated tests, this has not presented any noticeable problems. This is what the current PR does.
2. OR: use `update_module_enable()`. Notice that this function is a "light" enable. If this is the preferred way, we'll need to implement `hook_schema_0()` for modules that need to install a base schema before any of the hook_update_N are invoked. (NB: `update_module_enable()` doesn't invoke `backdrop_set_installed_schema_version()`, meaning that ALL hook_update_N hooks for the module WILL be called later in the process). Also, if this is the preferred way, we need to be sure that it calls `config_install_default_config()` to also install all the module's config files.

To me, number 1 above is the way to go, given that all tests I've done seem to indicate that `module_enable()` is not problematic in these cases. Feedback and thoughts?

**Comment by argiepiano** [comment]
@bugfolder - could you test and do a code review of the latest PR? We are losing momentum. I feel this is an important fix for a problem many D7 upgrades will face.

**Comment by bugfolder** [comment]
I was kinda hoping that folks with more experience in the innards of updating would  chime in, but time has passed, so here goes. If we do (2), then in addition to it being extra complication to manually install config files, all `hook_update_N()`s will be called for the module, which seems unnecessary since that wouldn't happen if the user had simply enabled the module manually prior to the upgrade (if that were possible, which, of course, it's not, which is why we're here).

So I'm in agreement: (1) is the way to go. Looking at the small number of `update_module_enable()` in core code, it looks like just a lower-level utility—lower level than what we need here.

**Comment by bugfolder** [comment]
A few small code comments, but otherwise looks good, code-wise. I'll run one more upgrade test.

**Comment by docwilmot** [comment]
Not an expert, but also agree with option 1. I've stared at the code and can't think of any complications right now.

**Comment by bugfolder** [comment]
I've run another test (upgrade with Rules and Ubercart). Worked like a charm, so testing-wise, LGTM. (Would still like to see the tiny code tweaks requested).

I'll advocate for this issue.

**Comment by argiepiano** [comment]
PR updated with suggestions. Thanks @bugfolder.

**Comment by bugfolder** [comment]
Code looks good!

**Comment by quicksketch** [comment]
Overall I think the concept looks good.

I found a few small issues in the PR: https://github.com/backdrop/backdrop/pull/3950#pullrequestreview-906728443

**Comment by quicksketch** [comment]
Added this to the 1.22.0 milestone and added @bugfolder as the advocate in the summary.

**Comment by quicksketch** [comment]
> Because modules like Redirect, Email and others must create either tables in the DB, or config files, the only possible way to enable them automatically is to use `module_enable()`, instead of the "lighter" `update_module_enable()`, which is typically used in update processes.

I don't think using `module_enable()` will work reliably for a couple of reasons:

* Firing module install hooks (like `hook_modules_installed()`) probably isn't safe at the point of a D7 database that hasn't yet had any updates applied. Modules will be expecting a fully updated database at the time they enable, not a Drupal 7 database structure (with virtually no configuration files yet created). Anything that tried to check for content types, fields, views, or pretty much anything in config would either throw an error or be unable to find anything.
* Using `module_install()` puts the module schema at the _latest_ version. So no update hooks can actually be run for that module. A module may need to do something different depending on whether it is upgraded from Drupal 7 version if it is installed fresh on Backdrop, such as copying settings or variables. And it may also need to interlace its updates before or after other update hooks.

The example of Redirect is actually a pretty good one to demonstrate. If a module declared Redirect as a dependency, it would be enabled prior to everything else via module_enable(). That would currently set its schema value to 1003 (based on `redirect_update_1003()`, and would cause it to skip `redirect_update_1000()`, which migrates the Drupal 7 settings into config. Instead Redirect module would be installed with _default_ settings, rather than the ones coming from D7.

I believe the intention of the current `update_module_enable()` is that if database tables are needed as part of an upgrade from Drupal 7, you'd created a `hook_schema_0()` function, and then any config would be added through `hook_update_1000()` (and subsequent numbers). Although looking around I see that `hook_schema_0()` isn't actually documented, even though it's been a hook since 2011.

**Comment by argiepiano** [comment]
I made the changes suggested by @quicksketch . Anything else we are waiting for?

**Comment by argiepiano** [comment]
OK, I just read @quicksketch's last comment. 

I'm stumped now. This is a catch 22. We can't use `module_enable` and we can't use `update_module_enable()` either as explained in the PR. The latter one will produce fatal errors since it doesn't actually load the `entity` module.

This means that we are back to square 1? I'd be happy to let others solve this, unless someone can point me in the right direction

**Comment by argiepiano** [comment]
> I believe the intention of the current update_module_enable() is that if database tables are needed as part of an upgrade from Drupal 7, you'd created a hook_schema_0() function, and then any config would be added through hook_update_1000()

So, as I see it, we are struggling with two different issues:

1. Enabling `entity` early on in the upgrade process to allow other required modules that need `entity` to be enabled. This could be done with `module_enable()` as in the current PR (in `update_upgrade_enable_dependencies()`, OR with a combination  of `update_module_enable('entity')` and `backdrop_load('entity')` to actually load the code and avoid a Function not found fatal error when trying to enable the required modules
2. Automatically enabling the Backdrop-only modules required by Backdrop's Standard installation (Redirect, Email, etc) and other required contrib modules (Entity Plus, Rules) during the upgrade. Using `update_module_enable()` for these is problematic (most glaringly, in the case of Redirect) as `update_module_enable()` does not created the necessary tables and config files. Thus it was necessary to use `module_enable()`, which made step 1 above necessary

Anyone can come up with a better way to do this?

**Comment by bugfolder** [comment]
@quicksketch wrote:

> Using module_install() puts the module schema at the latest version. So no update hooks can actually be run for that module. A module may need to do something different depending on whether it is upgraded from Drupal 7 version if it is installed fresh on Backdrop

But these modules that are being enabled via this special processing are modules _that didn't exist in Drupal D7_. They _should_ have the module schema at the latest version. In most cases, they should not have `hook_update_N()` functions (and the use cases that are driving this issue—Entity Plus, Entity UI, Ubercart—don't.)

So perhaps we could continue to use `module_enable()`, but also check all the modules that are going to be enabled at that time and whether they have any `hook_update_N()` implementations, and if they do, not attempt to enable them (and perhaps put up a warning message, "Sorry, bud, you're on your own to sort this one out.").

**Comment by argiepiano** [comment]
@bugfolder said:
> But these modules that are being enabled via this special processing are modules that didn't exist in Drupal D7. They should have the module schema at the latest version.

However, @quicksketch has a point. With Redirect, we would need to  redirect_update_1000 to run.  `module_enable()` won't run it:

@quicksketch  said:
> The example of Redirect is actually a pretty good one to demonstrate. If a module declared Redirect as a dependency, it would be enabled prior to everything else via module_enable(). That would currently set its schema value to 1003 (based on redirect_update_1003(), and would cause it to skip redirect_update_1000(), which migrates the Drupal 7 settings into config. Instead Redirect module would be installed with default settings, rather than the ones coming from D7.

**Comment by argiepiano** [comment]
So here's another thought: 

This thread was started with the intent to automatically enable required contrib modules that did not exist in D7 (specifically Entity Plus and and Entity UI).  We could go back to that **original intent** of this thread and provide a PR that only focuses on Entity Plus and Entity UI.  And leave alone all Backdrop core modules required by the Standard installation that are not currently enabled (which is the current state of affairs).

**Comment by bugfolder** [comment]
I wonder if there's some way within a module that it can indicate "I'm safe to auto-enable as part of a D7 upgrade," so that other modules (e.g., Ubercart) can take advantage of the option.

**Comment by argiepiano** [comment]
@bugfolder said: 

> I wonder if there's some way within a module that it can indicate "I'm safe to auto-enable as part of a D7 upgrade," so that other modules (e.g., Ubercart) can take advantage of the option.

You had proposed something along those lines (but not quite) earlier in this thread. You had suggested a new hook, `hook_declare_upgrade_required_modules()` to declare modules that needed to be enabled. Maybe worth revisiting this idea?

**Comment by argiepiano** [comment]
@bugfolder and @quicksketch: I've been doing more digging. Good news. Here's an update.

One of @quicksketch's concerns about using `module_enable()` during the upgrade "auto-enable" process (see [comment above](https://github.com/backdrop/backdrop-issues/issues/5499#issuecomment-1065829003)) was that, by using `module_update()` we would be skipping all `hook_update_N()` for the auto-enabled modules. @quicksketch offered as an example the module Redirect.  Redirect was contrib in D7, but is part of core, and required by the Standard Backdrop installation. Therefore it's one of the core modules automatically enabled during the new upgrade process from D7 provided by my PR.  

@quicksketch's concern was that, if we used `module_enable()` in the upgrade process to automatically enable Redirect, any D7 Redirect configuration would be lost, since Redirect's D7 variable configurations are converted into config (CMI) file configurations by `redirect_update_1000()`, and this hook is not triggered when using `module_enable()`.

However, @quicksketch, this is not completely correct - there is a twist.  I've done testing, and this is what happens. I'm offering two scenarios

Scenario 1: Your D7 site did not have Redirect enabled
----
In this case, there are no Redirect configurations in the D7 site, since it was not enabled. Therefore, it's safe to use `module_enable()` in the upgrade process. Using it will create the necessary table, and will install the default Backdrop configuration for Redirect. Nothing is lost.

Scenario 2: Your D7 site has Redirect enabled
----
When this is the case, Redirect is marked as "enabled" in the D7 system table, and therefore the code in my PR will not include Redirect as one of the modules to be automatically enabled. Instead, the "normal" upgrade workflow will take over: it will basically run all `redirect_update_N()` hooks normally (as it does without my PR) and it will convert all D7 variable configurations to Backdrop CMI.

So, `module_enable()` seems safe in these situations. Does this answer your concern? Is there anything else I'm missing? If not, it seems to me that this PR's approach is sensible and safe, and it doesn't need additional work. Please let me know if this is not the case.

P.S.
----   

The other core modules enabled automatically by the new upgrade process in a Standard installation are:
* Administration Bar - using `module_enable` is needed here since it needs to create a table. No "variable to config" conversion
* CKEditor - no schema or config conversion
* Configuration Manager - no schema 
* Date - `module_enable` needed to create schema tables if D7's contrib `date` was not installed
* Telemetry - no schema of configuration conversion
* Email - `module_enable` needed to create table if D7 contrib `email` was not installed
* Project Installer (!). No schema
* Layout (!). `module_enable` needed to create schema table
* Link. `module_enable` needed to create schema table
* Views UI. No schema. This module is in the same situation as Redirect: its `update_1000` hook moves D7 variable configuration to CMI. As explained above, this hook is not needed if your D7 site did not have Views UI enabled (no D7 configurations existed since it was not enabled). If your D7 site did have it enabled, then the new auto-enabling will not run for this module; instead the current upgrade workflow will be run, which will trigger `views_ui_update_1000()` to move the configurations.

**Comment by bugfolder** [comment]
This explanation seems pretty clear; thanks, @argiepiano, for explaining it! It would be get to get @quicksketch's concurrence, but then I think we can and should proceed  .

**Comment by argiepiano** [comment]
I have discover one problem with the current approach:

1. Start with a Standard D7 install
2. Disable Dashboard in the D7 site
3. Proceed to a Backdrop upgrade

During the upgrade process the following error occurs: `Error: Call to undefined function layout_reset_caches() in dashboard_install() (line 11 of`

I believe this is generated because we are auto-enabling Dashboard before Layout is enabled.  More work, testing and debugging needed....

**Comment by argiepiano** [comment]
After further consideration, and given the issue I discovered in the previous comment when using `module_enable()` I've decided to try a much "gentler" approach:
- Only required non-core, Backdrop-only modules are automatically enabled (e.g. Entity Plus)
- All enabling is now done with `update_module_enable()` which avoids triggering any hooks etc.

I've tested with D7 upgrades from a basic Standard install, and also with and without Ubercart and Rules.  Needs code review and further test, please..

**Comment by bugfolder** [comment]
Tested with a Rules upgrade (worked) and an Ubercart upgrade (no difference from same upgrade with stub modules). Will do some more testing tomorrow, but looking good!

**Comment by argiepiano** [comment]
Thanks @bugfolder. I just committed your docblock suggestion.

**Comment by argiepiano** [comment]
One aspect that may need to be documented somewhere is that, if your module depends on a Backdrop-only module that needs to create a table, the new approach in the PR will not create it by default (since it uses `update_module_enable()` . The module needs to implement `hook_schema_0()` which is invoked during `update_module_enable()`.

`hook_schema_0()` is not documented anywhere. I'll add it to this PR. It should probably be documented in `system.api.php`, right?

**Comment by bugfolder** [comment]
>  I'll add [documentation of `hook_schema_0()`] to this PR.

I think that should be a separate issue; since `hook_schema_0` is invoked by a public API function, it too is a public API function and should be documented, whether or not this particular issue calls for it. (And yes, `system.api.php`, where `hook_schema()` and `hook_schema_alter()` are documented, seems to be the right place.)

I would further suggest that a note be added to the documentation of `hook_schema()` pointing out the existence of `hook_schema_0()` and its possible need, but leave the primary discussion in the docblock of `hook_schema_0()`. Others might have further opinions about the appropriate way to document these two functions' relationship, but the appropriate place to discuss how they are documented would be in the new issue.

**Comment by quicksketch** [comment]
Thanks @argiepiano for the thorough explanation and testing of different situations! I left one more comment on the PR about the point at which we're enabling the new modules: https://github.com/backdrop/backdrop/pull/3950/files#r831343018

The separate issue for `hook_schema_0()` is a great idea, that way we'll be able to reference that issue in the future.

**Comment by argiepiano** [comment]
@quicksketch I'm pasting my response to your comment to give others a chance to chime in if they wish. 

@quicksketch said:
> This [`update_upgrade_enable_dependency()`] perhaps should be moved before the "Apply pending updates" button is clicked? The update_fix_compatibility() function does the exact opposite of this function: it disables old modules. This function enables new ones. Perhaps the two of them could be either named similar or be merged together some how?

> By moving up to above, before the updates are applied, any installation requirements that are not met by the newly installed modules will be displayed by update_check_requirements(). For example if a newly installed module required a PHP extension that we not available, that should be installed before attempting to upgrade the site, as update hooks may depend on the extension.

@quicksketch , I see the logic of your comment. One of the reason I decided to place this call later in the upgrade process was that the user would have a chance to see what modules would be automatically enabled before proceeding with the updates, giving them a chance to cancel. This was suggested in a comment by @docwilmot. If I move `update_upgrade_enable_dependency()` before `update_fix_compatibility()` then this option is gone.

If you are OK with this, I'll move it as requested. Let me know.  (BTW the call to `update_upgrade_enable_dependency()` could be moved up one step, to the operation stage  'selection' with the same result as the current placement, but I think you are asking me to move it even earlier, before the `update_fix_compatibility()` call)

**Comment by argiepiano** [comment]
@quicksketch  - any thoughts about my last question regarding moving this auto-enabling functionality earlier in the upgrade process?

**Comment by bugfolder** [comment]
Since this issue fell off the end of today's dev meeting and needs @quicksketch input: ping. 😉

**Comment by quicksketch** [comment]
@argiepiano Okay that makes sense. And the use-case of new dependencies not having requirements met is less likely than the need to inform users of new modules being enabled. I think that's fair.

In that case I think we may be ready to go on this one.

**Comment by quicksketch** [comment]
With no other concerns raised, I've merged https://github.com/backdrop/backdrop/pull/3950 into 1.x for 1.22.0. Thanks @argiepiano and @bugfolder for really paving the way here! Thanks @docwilmot and @klonos for input and reviews!
